setup.py

```diff
@@ -46,194 +46,332 @@
+            # Command to benefit from the same logging and dry_run logic as
+            # setuptools.
 
-        # In this example, we pass in the version to C++. You might not need to.
-        cmake_args += [f"-DEXAMPLE_VERSION_INFO={self.distribution.get_version()}"]
-
-        if self.compiler.compiler_type != "msvc":
-            # Using Ninja-build since it a) is available as a wheel and b)
-            # multithreads automatically. MSVC would require all variables be
-            # exported for Ninja to pick it up, which is a little tricky to do.
-            # Users can override the generator with CMAKE_GENERATOR in CMake
-            # 3.15+.
-            if not cmake_generator or cmake_generator == "Ninja":
-                try:
-                    import ninja
-
-                    ninja_executable_path = Path(ninja.BIN_DIR) / "ninja"
-                    cmake_args += [
-                        "-GNinja",
-                        f"-DCMAKE_MAKE_PROGRAM:FILEPATH={ninja_executable_path}",
-                    ]
-                except ImportError:
-                    pass
+            # Ensure that the destination directory exists.
+            self.mkpath(os.path.dirname(dst))
+            # Follow the example of the base build_py class by not preserving
+            # the mode. This ensures that the output file is read/write even if
+            # the input file is read-only.
+            self.copy_file(src, dst, preserve_mode=False)
 
-        else:
-            # Single config generators are handled "normally"
-            single_config = any(x in cmake_generator for x in {"NMake", "Ninja"})
 
-            # CMake allows an arch-in-generator style for backward compatibility
-            contains_arch = any(x in cmake_generator for x in {"ARM", "Win64"})
+# TODO(dbort): For editable wheels, may need to update get_source_files(),
+# get_outputs(), and get_output_mapping() to satisfy
+# https://setuptools.pypa.io/en/latest/userguide/extension.html#setuptools.command.build.SubCommand.get_output_mapping
 
-            # Specify the arch if using MSVC generator, but only if it doesn't
-            # contain a backward-compatibility arch spec already in the
-            # generator name.
-            if not single_config and not contains_arch:
-                cmake_args += ["-A", PLAT_TO_CMAKE[self.plat_name]]
 
-            # Multi-config generators have a different way to specify configs
-            if not single_config:
+class CustomBuild(build):
+    def initialize_options(self):
+        super().initialize_options()
+        # The default build_base directory is called "build", but we have a
+        # top-level directory with that name. Setting build_base in setup()
+        # doesn't affect this, so override the core build command.
+        #
+        # See build.initialize_options() in
+        # setuptools/_distutils/command/build.py for the default.
+        self.build_base = "pip-out"
```

Comment:
so one thing I noticed in the previous pip install experience, is that the build base directory name contains useful information such as python version and os arch. Do we still have that in the new directory name?

Reply:
Yeah, this just moves the root from `build/...` to `pip-out/...` so it doesn't sit on top of our existing `build/` directory. All of the platform-based directories live under this.